crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -523,31 +336,36 @@ unsafe impl WorldQuery for Entity {
     }
 }
 
-/// SAFETY: access is read only
-unsafe impl ReadOnlyWorldQuery for Entity {}
-
 /// SAFETY: `Self` is the same as `Self::ReadOnly`
-unsafe impl WorldQuery for EntityRef<'_> {
-    type Fetch<'w> = UnsafeWorldCell<'w>;
-    type Item<'w> = EntityRef<'w>;
+unsafe impl WorldQueryData for Entity {
     type ReadOnly = Self;
-    type State = ();
 
     fn shrink<'wlong: 'wshort, 'wshort>(item: Self::Item<'wlong>) -> Self::Item<'wshort> {
         item
     }
+}
 
-    const IS_DENSE: bool = true;
+/// SAFETY: access is read only
+unsafe impl ReadOnlyWorldQueryData for Entity {}
+
+/// SAFETY:
+/// `fetch` accesses all components in a readonly way.
+/// This is sound because `update_component_access` and `update_archetype_component_access` set read access for all components and panic when appropriate.
+/// Filters are unchanged.
+unsafe impl<'a> WorldQuery for EntityRef<'a> {
+    type Fetch<'w> = &'w World;
+    type Item<'w> = EntityRef<'w>;
+    type State = ();
 
-    const IS_ARCHETYPAL: bool = true;
+    const IS_DENSE: bool = true;
 
     unsafe fn init_fetch<'w>(
         world: UnsafeWorldCell<'w>,
         _state: &Self::State,
         _last_run: Tick,
         _this_run: Tick,
     ) -> Self::Fetch<'w> {
-        world
+        world.world()
```

Comment:
I believe at some point I needed to change the `Fetch` type of `EntityRef` to be `&'w World` instead of `UnsafeWorldCell<'w>` so I made this change (and the one below). I have just checked and that change is no longer necessary so I will revert.